### PR TITLE
Online Pertubation

### PR DESCRIPTION
**Pull request by cahuja1992**
Currently, Speed Perturbation has been implemented. As a lot of text is read in a very formal fashion, if someone speaks fast is not captured accurately. Thus, the speed_pertubation creates new samples with different rate of perturbations for new epochs.

### CODE REVIEW

**Comment by lissyx** [comment]
@cahuja1992 Thanks for that, but I don't think it's the right place to apply that kind of changes: deepspeech should remain focused on pulling the dataset, and the perturbations should be applied there, indeed creating a perturbated dataset.

**Comment by reuben** [comment]
@lissyx the idea here is to do online augmentation, where for every epoch slightly different perturbations are created, and hopefully achieving better generalization. Otherwise you'd have to save to disk every variation of the perturbed dataset before you even begin training.

**Comment by cahuja1992** [comment]
Okay, isn't it good that every training epoch gets to see different data so that the model becomes more generalized rather than getting overfit.

**Comment by reuben** [comment]
@cahuja1992 Thanks for the PR! Can you take a look at the linter check in the Travis job?

I see that pyrubberband is a wrapper over a command-line utility, so it creates temporary files and spawns a subprocess to process the audio. This could slow down training quite a bit, did you see a significant slowdown in your training runs? If there's no way to do the rubberbanding in memory, then at the very least we should do it before decoding the original WAV, since we're not using the original samples anyway.

Currently the flow looks like this: wav_filename -> read_file -> decode_wav -> rubberband (writes to a file, processes it into a new file, reads the new file) -> rest of input pipeline.

The initial file reading and WAV decoding are useless. We could just call rubberband on the original file, and read the modified file once.

File: util/feeding.py [hunk 1]
```diff
@@ -42,9 +43,12 @@ def samples_to_mfccs(samples, sample_rate):
 
 
 def audiofile_to_features(wav_filename):
+    global is_pertub
     samples = tf.io.read_file(wav_filename)
     decoded = contrib_audio.decode_wav(samples, desired_channels=1)
-    features, features_len = samples_to_mfccs(decoded.audio, decoded.sample_rate)
+    if is_pertub:
+        samples = pertub(decoded.audio)
```
Comment (alchemi5t) [inline util/feeding.py]: you're passing decoded audio to perturb and then perturb is trying to access 'audio' and 'sample_rate'attribute from there. Decoded audio is a numpy array. What's goin on here?

Also, if is_perturb==False, then shouldn't you be passing decoded.audio to samples_to_mfcc?

**Comment by lissyx** [comment]
> @lissyx the idea here is to do online augmentation, where for every epoch slightly different perturbations are created, and hopefully achieving better generalization. Otherwise you'd have to save to disk every variation of the perturbed dataset before you even begin training.

Sure, I'm just wondering / worrying about reproductibility. For some reason, I kept thinking in my mind that we wanted to avoid explicitely that kind of behavior.

File: util/feeding.py [hunk 2]
```diff
@@ -14,8 +14,9 @@
 
 from util.config import Config
 from util.text import text_to_char_array
+from pertub import speed_pertub, pertub
 
-
+is_pertub = False
```
Comment (lissyx) [inline util/feeding.py]: This should rather be exposed as a flag in `utils/flags.py`, defaulting to false as here.

**Comment by alchemi5t** [comment]
> worrying about reproductibility

How about using Random's RNG to get random int/float and set Random seed for reproducibility

File: util/feeding.py [hunk 3]
```diff
@@ -42,9 +43,12 @@ def samples_to_mfccs(samples, sample_rate):
 
 
 def audiofile_to_features(wav_filename):
+    global is_pertub
     samples = tf.io.read_file(wav_filename)
     decoded = contrib_audio.decode_wav(samples, desired_channels=1)
-    features, features_len = samples_to_mfccs(decoded.audio, decoded.sample_rate)
+    if is_pertub:
```
Comment (lissyx) [inline util/feeding.py]: if I read correctly, you're breaking the flow in case `is_perturb=False`, since we pass `samples` from `tf.io.read_file` to `samples_to_mfccs` instead of passing `decoded.audio`

File: requirements.txt [hunk 4]
```diff
@@ -20,4 +20,4 @@ soundfile
 # Miscellaneous scripts
 paramiko >= 2.1
 scipy
-matplotlib
+matplotlib
```
Comment (lissyx) [inline requirements.txt]: there's no change here

**Comment by alchemi5t** [comment]
@lissyx @cahuja1992 would another PR be warranted around this feature? i've added pitch modulation along with speed and randomized which augmentation at each call.

File: util/flags.py [hunk 5]
```diff
@@ -131,3 +131,6 @@ def create_flags():
                          lambda value: not value or os.path.isdir(value),
                          message='The path pointed to by --export_dir must exist and be a directory.')
 
+
+    # pertubation
+    f.DEFINE_boolean('pertubation', False, 'enable the online audio pertubation')
```
Comment (reuben) [inline util/flags.py]: This flag is never checked anywhere.

File: util/feeding.py [hunk 6]
```diff
@@ -44,7 +44,8 @@ def samples_to_mfccs(samples, sample_rate):
 def audiofile_to_features(wav_filename):
     samples = tf.io.read_file(wav_filename)
     decoded = contrib_audio.decode_wav(samples, desired_channels=1)
-    features, features_len = samples_to_mfccs(decoded.audio, decoded.sample_rate)
+    audio = pertub(decoded.audio)
```
Comment (reuben) [inline util/feeding.py]: This needs to be behind a check for the flag you added, but the check can't be here, it needs to be in the caller of `create_dataset` otherwise you'll also augment the validation and test sets.

File: util/pertubation.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import librosa
+import numpy as np
+
+def speed_pertub(samples):
+    samples = samples.flatten().astype('float32')
+    speed_range = np.random.uniform(low=0.8, high=1.4)
+    samples = librosa.effects.time_stretch(samples, rate=speed_range)
+    return samples.reshape([-1, 1])
+
+def pertub(samples):
```
Comment (reuben) [inline util/pertubation.py]: Please give this a more descriptive name like `apply_random_time_stretch`.

File: util/flags.py (hunk 5 above)
Comment (reuben) [reply util/flags.py]: Please give this flag a more descriptive name and description. Maybe `apply_time_stretch` for the name and describe it as `randomly time stretch training samples. Requires the Python librosa package to be installed`.

File: requirements.txt (hunk 4 above)
Comment (reuben) [reply requirements.txt]: This is removing the newline at the end of the file. Please add it back.

File: util/pertubation.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    samples = tf.py_func(speed_pertub,[samples], tf.float32)
```
Comment (reuben) [inline util/pertubation.py]: This should be `tf.compat.v1.py_func`. Convention in the repo is to import it as `tfv1`. See DeepSpeech.py for example.

**Comment by cahuja1992** [comment]
I have restructured the augmentation framework. Created augmentation as a separate module all together in the utils which has two types of augmentation techniques for audio augmentation and spectrogram augmentation. As of now not integrating the spec augment into feeding.py because it requires mel_filterbanks not `contrib_audio.audio_spectrogram and contrib_audio.mfcc`.

**Comment by cahuja1992** [comment]
Refer [Run instructions](https://github.com/cahuja1992/DeepSpeech/tree/pertub#Data-Augmentation-or-Pertubation)

File: util/feeding.py (hunk 2 above)
Comment (cahuja1992) [reply util/feeding.py]: Done

File: README.md [hunk 9]
```diff
@@ -381,6 +386,11 @@ python3 DeepSpeech.py --n_hidden 2048 --checkpoint_dir path/to/checkpoint/folder
 
 Note: the released models were trained with `--n_hidden 2048`, so you need to use that same value when initializing from the release models.
 
+## Data Augmentation or Pertubation
+
+Online Augmentation is a useful technique to for more generalization of machine learning models. Thus, a pre-processing pipeline with various augmentation techniques on audio data and also on spectrogram. 
+Full Documentation `util/augumentation/README.md`. By default augmentation is not enabled, to use this option use flag `--audio_augmentation_config_file="${pwd}/util/augmentation/conf/audio_augmentation.json` for audio augmentation and `--spectrum_augmentation_config_file="${pwd}/util/augmentation/conf/spectrum_augmentation.json`
```
Comment (reuben) [inline README.md]: nit: Full documentation at

pwd is a command, not an environment variable. "${pwd}" just expands to nothing, which creates a path that points to `/util/augmentation`. Did you test this? Also, you don't need to prepend `pwd` to paths like this, relative paths are relative to `pwd` by default anyway.

File: util/augmentation/README.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
+# Augmentation
+
+## Audio Augmentation
+Configuration file
+
+```{json}
```
Comment (reuben) [inline util/augmentation/README.md]: Proper way to specify the language for syntax highlighting is without `{}`, just ` ```json`

File: util/augmentation/README.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        "type": "speed_pertub",
+        "params": {"low_speed" : 0.8, "high_speed" : 1.4},
+        "rate": 0.5 # how many times speed_pertubation has to be applied
+    }
+]
+```
+
+Api Usage
+```{python}
```
Comment (reuben) [inline util/augmentation/README.md]: Same here.

File: util/augmentation/README.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+from .augmentation import AudioAugmentationPipeline
+audio_augmentor = AudioAugmentationPipeline(config_file)
+audio_augmentor.transform(samples, fs)
+```
+
+## Spectrogram Augmentation
+Configuration file
+
+```{json}
```
Comment (reuben) [inline util/augmentation/README.md]: And here.

File: util/augmentation/README.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+[ 
+    {
+        "type": "time_warp",
+        "params": {"W" : 80},
+        "rate": 0.5 # how many times speed_pertubation has to be applied
+    }
+]
+```
+
+```{python}
```
Comment (reuben) [inline util/augmentation/README.md]: And here.

File: util/augmentation/README.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+from .augmentation import  SpectrumAugmentationPipeline
+spectrum_augmentor = SpectrumAugmentationPipeline(config_file)
+spectrum_augmentor = mel
```
Comment (reuben) [inline util/augmentation/README.md]: What is this line supposed to do?

File: util/augmentation/audio_augmentation/augmentation.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+        :param samples:
+        :type: tensor
+        :param fs:
+        :type: tensor
+        """
+        processed_samples = samples
+        for augmentor, rate in zip(self._augmentors, self._rates):
+            if self._rng.uniform(0., 1.) < rate:
+                processed_samples = augmentor.transform(samples, fs)
```
Comment (reuben) [inline util/augmentation/audio_augmentation/augmentation.py]: If there are more than one transformations in the pipeline, this will only apply the last one, as you're transforming from `samples` every time. It should transform `processed_samples` instead.

File: util/augmentation/audio_augmentation/augmentation.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+        return processed_samples
+
+    def _parse_pipeline_from_json(self, config_json):
+        try:
+            with open(config_json, 'r') as config:
```
Comment (reuben) [inline util/augmentation/audio_augmentation/augmentation.py]: nit: `config` is a bad name, use `config_file`.

File: util/augmentation/audio_augmentation/augmentation.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+                configs = json.load(config)
+            augmentors = [self._get_augmentor_by_name(config["type"], config["params"]) for config in configs]
+            rates = [config["rate"] for config in configs]
+        except Exception as e:
+            raise ValueError("Error parsing the audio augmentation pipeline %s" % str(e))
```
Comment (reuben) [inline util/augmentation/audio_augmentation/augmentation.py]: What exceptions are you trying to catch here? The `JSONDecoder` exception? If so, use that specifier, instead of a plain `Exception`, as it is too broad.

File: util/augmentation/spectrum_augmentation/augmentation.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        try:
+            with open(config_json, 'r') as config:
+                configs = json.load(config)
+            augmentors = [self._get_augmentor_by_name(config["type"], config["params"]) for config in configs]
+            rates = [config["rate"] for config in configs]
+        except Exception as e:
+            raise ValueError("Error parsing the audio augmentation pipeline %s" % str(e))
+        return augmentors, rates
+
+    def _get_augmentor_by_name(self, name, params):
```
Comment (reuben) [inline util/augmentation/spectrum_augmentation/augmentation.py]: Most of this file/class is exactly the same as `util/augmentation/audio_augmentation/augmentation.py`. Please refactor the shared code into a single class.

File: util/augmentation/spectrum_augmentation/base.py [hunk 19]
```diff
@@ -0,0 +1,18 @@
+from abc import ABCMeta, abstractmethod
+
+class AugmentorBase(object):
```
Comment (reuben) [inline util/augmentation/spectrum_augmentation/base.py]: Same for this file/class.

File: util/config.py [hunk 20]
```diff
@@ -10,7 +10,7 @@
 from util.gpu import get_available_gpus
 from util.logging import log_error
 from util.text import Alphabet
-
+from util.augmentation import AudioAugmentationPipeline, SpectrumAugmentationPipeline
```
Comment (reuben) [inline util/config.py]: nit: two lines between end of imports and first class definition. Also, put the new import line in sorted order with the rest of the imports.

File: util/feeding.py [hunk 21]
```diff
@@ -15,7 +15,6 @@
 from util.config import Config
 from util.text import text_to_char_array
 
-
```
Comment (reuben) [inline util/feeding.py]: Undo this change.

File: util/feeding.py [hunk 22]
```diff
@@ -36,15 +35,19 @@ def samples_to_mfccs(samples, sample_rate):
                                                   stride=Config.audio_step_samples,
                                                   magnitude_squared=True)
     mfccs = contrib_audio.mfcc(spectrogram, sample_rate, dct_coefficient_count=Config.n_input)
+    #FIXME: w.r.t specAugment the audio_spectrogram needs to be changed to mel_frequence bank in log domain
+    # mfccs = Config.spectrum_augmentator.transform(mfccs) if Config.spectrum_augmentator else mfccs
     mfccs = tf.reshape(mfccs, [-1, Config.n_input])
-
+
```
Comment (reuben) [inline util/feeding.py]: nit: trailing whitespace

Comment (reuben) [reply util/feeding.py]: note to self: How is this passing the linter?

File: util/feeding.py [hunk 23]
```diff
@@ -36,15 +35,19 @@ def samples_to_mfccs(samples, sample_rate):
...
-
+    
     return mfccs, tf.shape(mfccs)[0]
 
 
 def audiofile_to_features(wav_filename):
     samples = tf.io.read_file(wav_filename)
     decoded = contrib_audio.decode_wav(samples, desired_channels=1)
-    features, features_len = samples_to_mfccs(decoded.audio, decoded.sample_rate)
+    audio = Config.audio_augmentator.transform(decoded.audio, decoded.sample_rate) if Config.audio_augmentator else decoded.audio
```
Comment (reuben) [inline util/feeding.py]: Use a normal if instead of an if expression here, as it's easier to read.

File: util/flags.py [hunk 24]
```diff
@@ -131,3 +131,8 @@ def create_flags():
                          lambda value: not value or os.path.isdir(value),
```
Comment (reuben) [inline util/flags.py]: Undo the executable bit that was set on this file.

File: util/pertubation.py (hunk 7 above)
Comment (reuben) [reply util/pertubation.py]: Seems like this file was leftover with your other changes and is no longer used.

**Comment by reuben** [comment]
Oh, something I forgot to mention. In this PR, everywhere that says "pertubation" should say "perturbation" (note the added 'r'). Including file names, documentation, etc.

File: README.md [hunk 25]
```diff
@@ -30,29 +30,34 @@ See the output of `deepspeech -h` for more information on the use of `deepspeech
...
-  - [Common Voice training data](#common-voice-training-data)
-  - [Training a model](#training-a-model)
-  - [Checkpointing](#checkpointing)
-  - [Exporting a model for inference](#exporting-a-model-for-inference)
-  - [Exporting a model for TFLite](#exporting-a-model-for-tflite)
-  - [Making a mmap-able model for inference](#making-a-mmap-able-model-for-inference)
-  - [Continuing training from a release model](#continuing-training-from-a-release-model)
-- [Contribution guidelines](#contribution-guidelines)
-- [Contact/Getting Help](#contactgetting-help)
+- [Project DeepSpeech](#Project-DeepSpeech)
```
[the remark this answers is not in the document]
Comment (cahuja1992) [reply README.md]: Should I add it towards the end?

[the remark this answers is not in the document]
Comment (reuben) [reply README.md]: Where you added it currently is fine, just without the extra level of indentation.

File: util/pertubation.py (hunk 7 above)
Comment (cahuja1992) [reply util/pertubation.py]: removed

**Comment by rhamnett** [comment]
It would be good to get this merged with the spelling mistake corrected as soon as poss

**Comment by cahuja1992** [comment]
Shall I reopen the new issue?

**Comment by rhamnett** [comment]
> Shall I reopen the new issue, with the corrected name ?

@cahuja1992 please do, this is amazing work, want to get it usable ASAP, read the Google info at https://ai.googleblog.com/2019/04/specaugment-new-data-augmentation.html - awesome
